lib/shell/run_python.sh

```diff
@@ -2,27 +2,67 @@
 run_python () {
-
+   trap - ERR # allow internal error handling
+   
     # get arguments
     program="$1"
     logfile="$2"
+    OUTPUT_DIR=$(dirname "$logfile")
 
     # set python command if unset
     if [ -z "$pythonCmd" ]; then
-        echo "No python command set. Using default: python"
+        echo -e "\nNo python command set. Using default: python"
         pythonCmd="python"
     fi
 
-    # run program, add output to logfile
-    echo "Executing: ${pythonCmd} ${program}"
-    (
-        ${pythonCmd} ${program} >> "${logfile}"
+    # check if the command exists before running, log error if does not
+    if ! command -v ${pythonCmd} &> /dev/null; then
+        error_time=$(date '+%Y-%m-%d %H:%M:%S')
+        echo -e "\033[0;31mProgram error\033[0m at ${error_time}: ${pythonCmd} not found. Make sure command line usage is properly set up." 
+        echo "Program Error at ${error_time}: ${pythonCmd} not found." >> "${logfile}"
+        return 1  # exit early with an error code
+    fi
+
+    # check if the target script exists
+    if [ ! -f "${program}" ]; then
+        error_time=$(date '+%Y-%m-%d %H:%M:%S')
+        echo -e "\033[0;31mProgram error\033[0m at ${error_time}: script ${program} not found." 
+        echo "Program Error at ${error_time}: script ${program} not found." >> "${logfile}"
+        return 1
+    fi
+
+    # capture the content of output folder before running the script
+    files_before=$(ls -1 "$OUTPUT_DIR" | grep -v "make.log" | tr '\n' ' ')
+
+    # log start time for the script
+    echo -e "\nScript ${program} in ${pythonCmd} started at $(date '+%Y-%m-%d %H:%M:%S')" | tee -a "${logfile}"
 
-        # report on errors
-        return_code=$?
-        if [ $return_code -ne 0 ]; then
-            # Log an error message if the command failed
-            echo "Error: ${program} failed with exit code $return_code"
+    # run command and capture both stdout and stderr in the output variable
+    output=$(${pythonCmd} -u "${program}" 2>&1)
+    return_code=$?  # capture the exit status 
+
+    # capture the content of output folder after running the script
+    files_after=$(ls -1 "$OUTPUT_DIR" | grep -v "make.log" | tr '\n' ' ')
+
+    # determine the new files that were created
+    created_files=$(comm -13 <(echo "$files_before") <(echo "$files_after"))
+
+    # report on errors or success and display the output
+    if [ $return_code -ne 0 ]; then
+        error_time=$(date '+%Y-%m-%d %H:%M:%S')
+        echo -e "\033[0;31mWarning\033[0m: ${program} failed at ${error_time}. Check log for details." # display error warning in terminal
+        echo "Error in ${program} at ${error_time}: $output" >> "${logfile}"  # log error output
+        if [ -n "$created_files" ]; then
+            echo -e "\033[0;31mWarning\033[0m: there was an error, but files where created. Check log." 
+            echo -e "\nWarning: There was an error, but these files were created: $created_files" >> "${logfile}"  # log created files
```

Comment:
You are right here @ShiqiYang2022 , but we need to think about what behavior we want. I suggest you look into that while testing the scripts and let me know what you think - see my reply in your previous [comment](https://github.com/gentzkow/GentzkowLabTemplate/pull/19/files#r1783127378).

Reply:
@lucamlouzada My worries is that we have nothing returned to `make.sh` here, not about choosing between `exit 1` and `return 1`. I think we at least need to return the value, either error or not error into `make.sh`. Does that make sense?

Reply:
Yes, I think you are right in terms of proper coding. Perhaps the correct approach is to propagate the error code whenever it happens (with no need to propagate 0 if it is successful). Do you agree? If so, I will implement this now before you start testing the scripts. @ShiqiYang2022

Reply:
@lucamlouzada Yes that's exactly what I meant, just add another `exit 1` or `return 1` in the corresponding script (and check all `run_xxx.sh`) to make sure it returns 1 to the `make.sh`!

Reply:
I have pushed these changes in 9e7d338. Now, the `run_xx` scripts will propagate the error status to `make.sh` as I added `exit 1 `to all the error clauses. The behavior of `exit 1 `vs `return 1` is not the same in this case (see [here](https://stackoverflow.com/questions/4419952/difference-between-return-and-exit-in-bash-functions) for a great explanation), but I conducted some tests and believe` exit 1` is better since we are running these scripts inside a subshell in `make.sh`. This also increases the consistency between `zsh` and `bash`: now, from what I have tested, the difference is that in `bash` the error in `run_xx `causes `make.sh` to exit the subshell, without triggering the `make.sh` error trap, while in `zsh` it exits the subshell **and** triggers the `make.sh` error trap. If we used `return 1` instead, the parent `make.sh` does not detect the error and does not leave the subshell.  I believe this is now ready for you to test @ShiqiYang2022 . Thanks again for all your help and feedback